src/kafka-util/src/client.rs

```diff
@@ -625,16 +623,105 @@ pub fn get_partitions<C: ClientContext>(
     Ok(partition_ids)
 }
 
+/// Default to true as they hey have no downsides <https://github.com/confluentinc/librdkafka/issues/283>.
+pub const DEFAULT_KEEPALIVE: bool = true;
+/// The `rdkafka` default.
+/// - <https://github.com/confluentinc/librdkafka/blob/master/CONFIGURATION.md>
+pub const DEFAULT_SOCKET_TIMEOUT: Duration = Duration::from_millis(60000);
+/// The `rdkafka` default.
+/// - <https://github.com/confluentinc/librdkafka/blob/master/CONFIGURATION.md>
+pub const DEFAULT_SOCKET_CONNECTION_SETUP_TIMEOUT: Duration = Duration::from_millis(30000);
+
+/// A reasonable default timeout when fetching metadata or partitions.
+pub const DEFAULT_FETCH_METADATA_TIMEOUT: Duration = Duration::from_secs(10);
+
+/// Configurable timeouts for Kafka connections.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct TcpTimeoutConfig {
+    /// Whether or not to enable
+    pub keepalive: bool,
+    /// The timeout for network requests.
+    pub socket_timeout: Duration,
+    /// The timeout for setting up network connections.
+    pub socket_connection_setup_timeout: Duration,
+    /// The timeout for fetching metadata from upstream.
+    pub fetch_metadata_timeout: Duration,
+}
+
+impl Default for TcpTimeoutConfig {
+    fn default() -> Self {
+        TcpTimeoutConfig {
+            keepalive: DEFAULT_KEEPALIVE,
+            socket_timeout: DEFAULT_SOCKET_TIMEOUT,
+            socket_connection_setup_timeout: DEFAULT_SOCKET_CONNECTION_SETUP_TIMEOUT,
+            fetch_metadata_timeout: DEFAULT_FETCH_METADATA_TIMEOUT,
+        }
+    }
+}
+
+impl TcpTimeoutConfig {
+    /// Build a `TcpTimeoutConfig` from the given parameters. Parameters outside the supported
+    /// range are defaulted and cause an error log.
+    pub fn build(
+        keepalive: bool,
+        socket_timeout: Duration,
+        socket_connection_setup_timeout: Duration,
+        fetch_metadata_timeout: Duration,
+    ) -> TcpTimeoutConfig {
+        // Constrain values based on ranges here:
+        // <https://github.com/confluentinc/librdkafka/blob/master/CONFIGURATION.md>
+        //
+        // Note we error log but do not fail as this is called in a non-fallible
+        // LD-sync in the adapter.
+
+        // The documented max here is `300000`, but rdkafka bans `socket.timeout.ms` being more
+        // than `transaction.timeout.ms` + 100ms. We don't currently allow configuring
+        // `transaction.timeout.ms`, so we use its default + 100 here.
+        let socket_timeout = if socket_timeout.as_millis() > 60100 {
```

Comment:
Can we just do all these checks when we set the timeout rather than at runtime? Logging these as errors is going to fire sentry events, which seems like an odd interaction. e.g. look at `ValueConstraint::Domain`

Reply:
til about `ValueConstraint::Domain`, but I am about to add `transaction.timeout.ms` support which bounds the acceptable values for `socket.timeout.ms`, so there is a relationship that can't be encoded there :(